linalg: Eigenvalues and Eigenvectors (#816)
fortran-lang/stdlib: doc/specs/stdlib_linalg.md

```fortran
{!example/linalg/example_determinant2.f90!}
```

## `eig` - Eigenvalues and Eigenvectors of a Square Matrix

### Status

Experimental

### Description

This subroutine computes the solution to the eigenproblem \( A \cdot \bar{v} - \lambda \cdot \bar{v} \), where \( A \) is a square, full-rank, `real` or `complex` matrix.

Result array `lambda` returns the eigenvalues of \( A \). The user can request eigenvectors to be returned: 

> on output `left` may contain the left eigenvectors, `right` the right eigenvectors of \( A \). 

Reviewer: This sentence is unclear. Is it something missing after "returned:"?

Author: Thank you @jvdp1: I've refactored it here [f13e75d](https://github.com/fortran-lang/stdlib/pull/816/commits/f13e75d76f12af4c2084efc0b962d05b1474ebc2)

> Each eigenvector is returned as a column of `left` of `right`.

Reviewer: ```suggestion
Each eigenvector is returned as a column of `left` or `right`.
```

Author: Also [f13e75d](https://github.com/fortran-lang/stdlib/pull/816/commits/f13e75d76f12af4c2084efc0b962d05b1474ebc2)

The solver is based on LAPACK's `*GEEV` backends.

### Syntax

`call ` [[stdlib_linalg(module):eig(interface)]] `(a, lambda [, right] [,left] [,overwrite_a] [,err])`

### Arguments

> `a` : `real` or `complex` square array containing the coefficient matrix. It is normally an `intent(in)` argument. If `overwrite_a=.true.`, it is an `intent(inout)` argument and is destroyed by the call. 

Reviewer: ```suggestion
`a` : `real` or `complex` square array containing the coefficient matrix. If `overwrite_a=.false.`, it is an `intent(in)` argument. Otherwise, it is an `intent(inout)` argument and is destroyed by the call. 
```

> `lambda`: Shall be a `complex` rank-1 array of the same kind as `a`, containing the eigenvalues. It is an `intent(out)` argument. 

Reviewer: Only `complex`? Or of the same type as `a`?
```suggestion
`lambda`: Shall be a `real` or `complex` rank-1 array of the same kind as `a`, containing the eigenvalues. It is an `intent(out)` argument. 
```

Author: See comment below: I've added the option to return the `real` part only (otherwise, it was `complex` only)

> `right` (optional): Shall be a `complex` rank-2 array of the same size and kind as `a`, containing the right eigenvectors of `a`. It is an `intent(out)` argument.

Reviewer: ```suggestion
`right` (optional): Shall be a `real` or `complex` rank-2 array of the same size and kind as `a`, containing the right eigenvectors of `a`. It is an `intent(out)` argument.
```

Author: See comment below: I've added the option to return the real part only (otherwise, it was complex only)

> `left` (optional): Shall be a `complex` rank-2 array of the same size and kind as `a`, containing the left eigenvectors of `a`. It is an `intent(out)` argument.

Reviewer: ```suggestion
`left` (optional): Shall be a `real` or `complex` rank-2 array of the same size and kind as `a`, containing the left eigenvectors of `a`. It is an `intent(out)` argument.
```

> `overwrite_a` (optional): Shall be an input logical flag. if `.true.`, input matrix `a` will be used as temporary storage and overwritten. This avoids internal data allocation. This is an `intent(in)` argument.

Reviewer: ```suggestion
`overwrite_a` (optional): Shall be an input `logical` flag. If `.true.`, the matrix `a` will be used as temporary storage and overwritten. This avoids internal data allocation. This is an `intent(in)` argument.
```
What is the default value (.true. or .false.)?

`err` (optional): Shall be a `type(linalg_state_type)` value. This is an `intent(out)` argument.

### Return value

Raises `LINALG_ERROR` if the calculation did not converge.
Raises `LINALG_VALUE_ERROR` if any matrix or arrays have invalid/incompatible sizes.
If `err` is not present, exceptions trigger an `error stop`.

### Example

```fortran
{!example/linalg/example_eig.f90!}
```

> ## `eigh` - Eigenvalues and Eigenvectors of a Real symmetric or Complex Hermitian Square Matrix

Reviewer: ```suggestion
## `eigh` - Eigenvalues and Eigenvectors of a Real Symmetric or Complex Hermitian Square Matrix
```

### Status

Experimental

### Description

This subroutine computes the solution to the eigendecomposition \( A \cdot \bar{v} - \lambda \cdot \bar{v} \),
where \( A \) is a square, full-rank, `real` symmetric \( A = A^T \) or `complex` Hermitian \( A = A^H \) matrix.

Result array `lambda` returns the `real` eigenvalues of \( A \). The user can request the orthogonal eigenvectors 

> to be returned: on output `vectors` may contain the matrix of eigenvectors, returned as a columns.

Reviewer: this sentence is not clear.

> Normally, only the lower triangular part of \( A \) is accessed. On input, `logical` flag `upper_a` 

Reviewer: ```suggestion
As default, only the lower triangular part of \( A \) is accessed. On input, `logical` flag `upper_a` 
```

allows the user to request what triangular part of the matrix should be used.

The solver is based on LAPACK's `*SYEV` and `*HEEV` backends.

### Syntax

`call ` [[stdlib_linalg(module):eigh(interface)]] `(a, lambda [, vectors] [, upper_a] [, overwrite_a] [,err])`

### Arguments

> `a` : `real` or `complex` square array containing the coefficient matrix. It is normally an `intent(in)` argument. If `overwrite_a=.true.`, it is an `intent(inout)` argument and is destroyed by the call. 

Reviewer: ```suggestion
`a` : `real` or `complex` square array containing the coefficient matrix. It is an `intent(in)` argument. If `overwrite_a=.true.`, it is an `intent(inout)` argument and is destroyed by the call. 
```

> `lambda`: Shall be a `complex` rank-1 array of the same kind as `a`, containing the eigenvalues. It is an `intent(out)` argument. 

Reviewer: ```suggestion
`lambda`: Shall be a rank-1 array of the same kind as `a`, containing the eigenvalues. It is an `intent(out)` argument. 
```

`vectors` (optional): Shall be a rank-2 array of the same type, size and kind as `a`, containing the eigenvectors of `a`. It is an `intent(out)` argument.

> `upper_a` (optional): Shall be an input logical flag. if `.true.`, the upper triangular part of `a` will be used accessed. Otherwise, the lower triangular part will be accessed. It is an `intent(in)` argument.

Reviewer: ```suggestion
`upper_a` (optional): Shall be an input `logical` flag. if `.true.`, the upper triangular part of `a` will be accessed. Otherwise, the lower triangular part will be accessed. It is an `intent(in)` argument.
```

> `overwrite_a` (optional): Shall be an input logical flag. if `.true.`, input matrix `a` will be used as temporary storage and overwritten. This avoids internal data allocation. This is an `intent(in)` argument.

Reviewer: ```suggestion
`overwrite_a` (optional): Shall be an input `logical` flag. If `.true.`, input matrix `a` will be used as temporary storage and overwritten. This avoids internal data allocation. This is an `intent(in)` argument.
```
What is the default?

`err` (optional): Shall be a `type(linalg_state_type)` value. This is an `intent(out)` argument.

### Return value

Raises `LINALG_ERROR` if the calculation did not converge.
Raises `LINALG_VALUE_ERROR` if any matrix or arrays have invalid/incompatible sizes.
Raises `LINALG_VALUE_ERROR` if the `real` component is only requested, but the eigenvalues have non-trivial imaginary parts.
If `err` is not present, exceptions trigger an `error stop`.

### Example

```fortran
{!example/linalg/example_eig.f90!}
```

## `eigh` - Eigenvalues and Eigenvectors of a Real symmetric or Complex Hermitian Square Matrix

### Status

Experimental

### Description

This subroutine computes the solution to the eigendecomposition \( A \cdot \bar{v} - \lambda \cdot \bar{v} \),
where \( A \) is a square, full-rank, `real` symmetric \( A = A^T \) or `complex` Hermitian \( A = A^H \) matrix.

Result array `lambda` returns the `real` eigenvalues of \( A \). The user can request the orthogonal eigenvectors 

> to be returned: on output `vectors` may contain the matrix of eigenvectors, returned as a columns.

Reviewer: ```suggestion
to be returned: on output `vectors` may contain the matrix of eigenvectors, returned as a column.
```

Normally, only the lower triangular part of \( A \) is accessed. On input, `logical` flag `upper_a` 
allows the user to request what triangular part of the matrix should be used.

The solver is based on LAPACK's `*SYEV` and `*HEEV` backends.

### Syntax

`call ` [[stdlib_linalg(module):eigh(interface)]] `(a, lambda [, vectors] [, upper_a] [, overwrite_a] [,err])`

### Arguments

`a` : `real` or `complex` square array containing the coefficient matrix. It is an `intent(in)` argument. If `overwrite_a=.true.`, it is an `intent(inout)` argument and is destroyed by the call. 

`lambda`: Shall be a `complex` rank-1 array of the same precision as `a`, containing the eigenvalues. It is an `intent(out)` argument. 

`vectors` (optional): Shall be a rank-2 array of the same type, size and kind as `a`, containing the eigenvectors of `a`. It is an `intent(out)` argument.

> `upper_a` (optional): Shall be an input `logical` flag. if `.true.`, the upper triangular part of `a` will be accessed. Otherwise, the lower triangular part will be accessed. It is an `intent(in)` argument.

Reviewer: ```suggestion
`upper_a` (optional): Shall be an input `logical` flag. If `.true.`, the upper triangular part of `a` will be accessed. Otherwise, the lower triangular part will be accessed. It is an `intent(in)` argument.
```

`overwrite_a` (optional): Shall be an input `logical` flag. If `.true.`, input matrix `a` will be used as temporary storage and overwritten. This avoids internal data allocation. This is an `intent(in)` argument.

`err` (optional): Shall be a `type(linalg_state_type)` value. This is an `intent(out)` argument.

### Return value

Raises `LINALG_ERROR` if the calculation did not converge.
Raises `LINALG_VALUE_ERROR` if any matrix or arrays have invalid/incompatible sizes.
If `err` is not present, exceptions trigger an `error stop`.

### Example

```fortran
{!example/linalg/example_eigh.f90!}
```

## `eigvals` - Eigenvalues of a Square Matrix

### Status

Experimental

### Description

This function returns the eigenvalues to matrix \( A \): a square, full-rank, `real` or `complex` matrix.
The eigenvalues are solutions to the eigenproblem \( A \cdot \bar{v} - \lambda \cdot \bar{v} \).

Result array `lambda` is `complex`, and returns the eigenvalues of \( A \). 
The solver is based on LAPACK's `*GEEV` backends.

### Syntax

`lambda = ` [[stdlib_linalg(module):eigvals(interface)]] `(a, [,err])`

### Arguments

> `a` : `real` or `complex` square array containing the coefficient matrix. It is normally an `intent(in)` argument.

Reviewer: ```suggestion
`a` : `real` or `complex` square array containing the coefficient matrix. It is an `intent(in)` argument.
```

`err` (optional): Shall be a `type(linalg_state_type)` value. This is an `intent(out)` argument.

### Return value

Returns a `complex` array containing the eigenvalues of `a`. 

Raises `LINALG_ERROR` if the calculation did not converge.
Raises `LINALG_VALUE_ERROR` if any matrix or arrays have invalid/incompatible sizes.
If `err` is not present, exceptions trigger an `error stop`.

### Example

```fortran
{!example/linalg/example_eigvals.f90!}
```

## `eigvalsh` - Eigenvalues of a Real Symmetric or Complex Hermitian Square Matrix

### Status

Experimental

### Description

This function returns the eigenvalues to matrix \( A \): a where \( A \) is a square, full-rank, 
`real` symmetric \( A = A^T \) or `complex` Hermitian \( A = A^H \) matrix.
The eigenvalues are solutions to the eigenproblem \( A \cdot \bar{v} - \lambda \cdot \bar{v} \).

Result array `lambda` is `real`, and returns the eigenvalues of \( A \). 
The solver is based on LAPACK's `*SYEV` and `*HEEV` backends.

### Syntax

`lambda = ` [[stdlib_linalg(module):eigvalsh(interface)]] `(a, [, upper_a] [,err])`

### Arguments

> `a` : `real` or `complex` square array containing the coefficient matrix. It is normally an `intent(in)` argument.

Reviewer: ```suggestion
`a` : `real` or `complex` square array containing the coefficient matrix. It is an `intent(in)` argument.
```

> `upper_a` (optional): Shall be an input logical flag. if `.true.`, the upper triangular part of `a` will be used accessed. Otherwise, the lower triangular part will be accessed (default). It is an `intent(in)` argument.

Reviewer: ```suggestion
`upper_a` (optional): Shall be an input logical flag. If `.true.`, the upper triangular part of `a` will be used accessed. Otherwise, the lower triangular part will be accessed (default). It is an `intent(in)` argument.
```
